onnxruntime/core/providers/qnn/builder/opbuilder/matmul_op_builder.cc

```diff
@@ -0,0 +1,228 @@
+    } else {
+      ORT_RETURN_IF_ERROR(qnn_model_wrapper.AddReshapeNode(org_input_1_name, input_1_name, input_info_1.shape, shape_2d,
+                                                           input_info_1.qnn_data_type, input_info_1.quant_param,
+                                                           quant_param_2d, do_op_validation,
+                                                           qnn_model_wrapper.IsGraphInput(org_input_1_name), false));
+    }
+  } else {
+    if (qnn_model_wrapper.IsQnnTensorWrapperExist(input_1_name)) {
+      LOGS(logger, VERBOSE) << "Tensor already added, skip it: " << input_1_name;
+    } else {
+      QnnTensorWrapper input_1_tensor;
+      ORT_RETURN_IF_ERROR(qnn_model_wrapper.MakeTensorWrapper(inputs[1], input_1_tensor));
+      ORT_RETURN_IF_NOT(qnn_model_wrapper.AddTensorWrapper(std::move(input_1_tensor)), "Failed to add tensor.");
+    }
+  }
+  input_names.emplace_back(input_1_name);
+
+  return Status::OK();
+}
+
+Status MatMulOpBuilder::ProcessAttributesAndOutputs(QnnModelWrapper& qnn_model_wrapper, const NodeUnit& node_unit,
+                                                    std::vector<std::string>&& input_names,
+                                                    [[maybe_unused]] const logging::Logger& logger,
+                                                    bool do_op_validation) const {
+  const auto& inputs = node_unit.Inputs();
+  TensorInfo input_info_0{};
+  TensorInfo input_info_1{};
+  bool use_fully_connected = false;
+  ORT_RETURN_IF_ERROR(
+      CheckInputs(qnn_model_wrapper, inputs[0], inputs[1], input_info_0, input_info_1, use_fully_connected));
+  bool reshape_input_0 = input_info_0.shape.size() == 1;
+  bool reshape_input_1 = input_info_1.shape.size() == 1;
+  bool reshape_output = input_info_0.shape.size() == 1 || input_info_1.shape.size() == 1 ||
+                        (use_fully_connected && input_info_0.shape.size() > 2);
+
+  const std::string& org_output_name = node_unit.Outputs()[0].node_arg.Name();
+  std::string op_output_name = org_output_name;
+  TensorInfo output_info{};
+  ORT_RETURN_IF_ERROR(qnn_model_wrapper.GetTensorInfo(node_unit.Outputs()[0], output_info));
+  std::vector<uint32_t> op_output_shape = output_info.shape;
+  QnnQuantParamsWrapper op_output_quant_param = output_info.quant_param.Copy();
+  if (reshape_output) {
+    op_output_name = org_output_name + "_ort_qnn_ep_reshape";
+    if (use_fully_connected && input_info_0.shape.size() > 2) {
+      op_output_shape = {std::accumulate(input_info_0.shape.begin(), input_info_0.shape.end() - 1,
+                                         static_cast<uint32_t>(1), std::multiplies<uint32_t>()),
+                         reshape_input_1 ? 1 : input_info_1.shape.back()};
+      ORT_ENFORCE(!op_output_quant_param.IsPerChannel());
+    } else {
+      // If both inputs are 1D tensors, the output shape is [1] instead of scalar.
+      if (reshape_input_1) {
```

Comment:
The reshape_input_0 is checked by following "else if". I thought if both side are 1D, the output is scalar, so we need to two "1". But actually it's [1], so only one "1" will be added, that's why "else if" is used.

Reply:
Ah ok. I was confused by the comment. Would something like this be clearer?

```
      // If input 1 is 1D we append 1 to the output shape.
      // If both inputs are 1D the output shape is [1] instead of scalar, so we also need to append 1 to the
      // output shape.
```

If it was me I'd also make the condition `if (reshape_input_1 || (reshape_input_0 && reshape_input_1))`. It costs next to nothing but makes the intention of the code explicit.